src/sentry/static/sentry/app/components/events/groupingInfo.jsx

```diff
@@ -0,0 +1,335 @@
+  padding: 15px 0 20px 0;
+  margin-top: 15px;
+  border-top: 1px solid ${p => p.theme.borderLighter};
+  ${p => (p.contributes ? '' : 'color:' + p.theme.gray6)};
+`;
+
+const StyledGroupVariantTitle = styled('h5')`
+  margin: 0 0 10px 0;
+  color: inherit !important;
+  text-transform: uppercase;
+  font-size: 14px;
+`;
+
+const StyledGroupingComponentWrapper = styled('div')`
+  border-top: 1px solid ${p => p.theme.borderLighter};
+  padding: 10px 0 0 0;
+  margin-top: -10px;
+`;
+
+const StyledGroupingComponentList = styled('ul')`
+  padding: 0;
+  margin: 0;
+  list-style: none;
+`;
+
+const StyledGroupingComponentListItem = styled('li')`
+  padding: 0;
+  margin: 2px 0 1px 13px;
+`;
+
+const StyledGroupingComponent = styled(({contributes, ...props}) => <div {...props} />)`
+  ${p => (p.contributes ? '' : 'color:' + p.theme.gray6)};
+`;
+
+const StyledGroupingValue = styled('code')`
+  display: inline-block;
+  margin: 1px 4px 1px 0;
+  font-size: 12px;
+  padding: 1px 2px;
+  color: inherit;
+`;
+
+class GroupingComponent extends React.Component {
+  static propTypes = {
+    component: PropTypes.object,
+    showNonContributing: PropTypes.bool,
+  };
+
+  render() {
+    const {component} = this.props;
+
+    const children = component.values.map((value, idx) => {
+      let rv;
+      if (_.isObject(value)) {
+        // no point rendering such nodes at all, we never show them
+        if (!value.contributes && !value.hint && value.values.length === 0) {
+          return null;
+        }
+        // non contributing values are otherwise optional
+        if (!this.props.showNonContributing && !value.contributes) {
+          return null;
+        }
+        rv = (
+          <GroupingComponent
+            component={value}
+            showNonContributing={this.props.showNonContributing}
+          />
+        );
+      } else {
+        rv = <StyledGroupingValue>{JSON.stringify(value, null, 2)}</StyledGroupingValue>;
+      }
+      return (
+        <StyledGroupingComponentListItem key={idx}>{rv}</StyledGroupingComponentListItem>
+      );
+    });
+
+    return (
+      <StyledGroupingComponent contributes={component.contributes}>
+        <span>
+          {component.name || component.id}
+          {component.hint && <small>{` (${component.hint})`}</small>}
+        </span>
+        <StyledGroupingComponentList>{children}</StyledGroupingComponentList>
+      </StyledGroupingComponent>
+    );
+  }
+}
+
+function hasNonContributingComponent(component) {
+  if (!component.contributes) {
+    return true;
+  }
+  for (const value of component.values) {
+    if (_.isObject(value) && hasNonContributingComponent(value)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+class GroupVariant extends React.Component {
+  static propTypes = {
+    variant: PropTypes.object,
+  };
+
+  constructor(...args) {
+    super(...args);
+    this.state = {
+      showNonContributing: false,
+    };
+  }
+
+  toggleNonContributing() {
```

Comment:
Just making lots of little suggestions here.

You can write this as `toggleNonContributing = () => {` and it will be a property of the instance, thus you can use `onClick={this.toggleNonContributing}` and you won't need to bind `this`.

Reply:
Changed.